### PR TITLE
Add abc decorators to DshResourceProvider

### LINKED ISSUE #1516: Apply abc decorators to DshResourceProvider
## :white_check_mark: Checklist

- [x] I have searched open and closed issues for duplicates.
- [x] This is a request for a new feature in the Data Safe Haven or an upgrade to an existing feature.
- [x] The feature is still missing in the [latest version](https://github.com/alan-turing-institute/data-safe-haven/releases).
- [x] I have read through the [documentation](https://alan-turing-institute.github.io/data-safe-haven/).
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :strawberry: Suggested change

- `DshResourceProvider` contains inherited methods that must be implemented by child classes
- This should be done using `abc` decorators

## :steam_locomotive: How could this be done?

Replace existing `NotImplementedError` with `@abstractmethod` as appropriate

### PR DESCRIPTION
**Pull request by jemrobinson**
## :white_check_mark: Checklist

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).

### :vertical_traffic_light: Depends on

n/a

### :arrow_heading_up: Summary

- Add abc decorators to DshResourceProvider
- Change logic about how `check()` is handled

### :closed_umbrella: Related issues

Closes #1516

### :microscope: Tests

Not tested

### CODE REVIEW

File: data_safe_haven/infrastructure/components/dynamic/blob_container_acl.py [hunk 1]
```diff
@@ -97,6 +97,10 @@ def diff(
         id(id_)
         return self.partial_diff(old_props, new_props)
 
+    def refresh(self, props: dict[str, Any]) -> dict[str, Any]:
+        """TODO: check whether ACLs have changed"""
```
Comment (JimMadge) [inline data_safe_haven/infrastructure/components/dynamic/blob_container_acl.py:101]: Should we open an issue for this?
I guess it is a known limitation of our implementation?

File: data_safe_haven/infrastructure/components/dynamic/file_upload.py [hunk 2]
```diff
@@ -117,6 +117,10 @@ def diff(
             )
         return self.partial_diff(old_props, new_props, [])
 
+    def refresh(self, props: dict[str, Any]) -> dict[str, Any]:
+        """TODO: check whether file has changed"""
```
Comment (JimMadge) [inline data_safe_haven/infrastructure/components/dynamic/file_upload.py:121]: Same as above

**Review by JimMadge — APPROVED** [review]
LGTM.

Do we have issues for the limited (or technically unimplemented ?) methods in concrete classes?

File: data_safe_haven/infrastructure/components/dynamic/blob_container_acl.py (hunk 1 above)
Comment (jemrobinson) [reply data_safe_haven/infrastructure/components/dynamic/blob_container_acl.py:101]: I think this code shouldn't be too hard to write but it's not relevant to this PR.

File: data_safe_haven/infrastructure/components/dynamic/file_upload.py (hunk 2 above)
Comment (jemrobinson) [reply data_safe_haven/infrastructure/components/dynamic/file_upload.py:121]: As above - I think it shouldn't be too hard to do, but isn't the focus of this PR.

**Comment by jemrobinson** [comment]
> Do we have issues for the limited (or technically unimplemented ?) methods in concrete classes?

Not sure I understand what you mean here but static type-checking should be catching any non-implemented methods.

**Pull request merged**